### PR TITLE
Add probe type to the Beam model

### PR DESCRIPTION
**Pull request by dagewa**
Fixes [#575](https://github.com/cctbx/dxtbx/issues/575).

This PR adds a `Probe` `enum` to `Beam`. In future this could be used to alter behaviour of algorithms, such as the QE and parallax corrections. At the moment this just adds the additional datum and allows serialisation etc. Old experiment files where this is not set are interpreted as having `Probe.xray`. Some electron diffraction format classes are updated here to record `Probe.electron`.

A DIALS PR will be required following this. Currently the only DIALS tests that fail are those for `dials.show`.

### CODE REVIEW

**Review by toastisme — APPROVED** [review]
This all looks fine to me, and should be quite useful.

**Comment by dagewa** [comment]
@toastisme I've made the changes to hopefully get `Probe` working with the `PolychromaticBeam`. Might have missed something, so if you have anything else you can check on your fork that would be good. One thing I wasn't sure of were the type hints for a value like `Probe.xray`, so I just left that blank.

**Comment by toastisme** [comment]
> @toastisme I've made the changes to hopefully get `Probe` working with the `PolychromaticBeam`. Might have missed something, so if you have anything else you can check on your fork that would be good. One thing I wasn't sure of were the type hints for a value like `Probe.xray`, so I just left that blank.

Had a look and looks good. I realised `Probe` wasn't part of the phil params so have added it [here](https://github.com/cctbx/dxtbx/pull/654), along with a suggestion to have the probe be the NeXus name in .expt files, rather than the enum int, to make it human readable.

File: src/dxtbx/model/beam.h [hunk 1]
```diff
@@ -289,6 +304,45 @@ namespace dxtbx { namespace model {
...
+      // https://manual.nexusformat.org/classes/base_classes/NXsource.html
+      switch (probe_) {
+      case xray:
+        return std::string("x-ray");
+      case electron:
+        return std::string("electron");
+      case neutron:
+        return std::string("neutron");
+      default:
+        return std::string("unknown");
```
Comment (ndevenish) [inline src/dxtbx/model/beam.h:322]:
```suggestion
        DXTBX_ERROR("Unknown probe type");
```

File: src/dxtbx/model/beam.h [hunk 2]
```diff
@@ -289,6 +304,45 @@ namespace dxtbx { namespace model {
...
+      if (probe == "x-ray") {
+        return Probe::xray;
+      } else if (probe == "electron") {
+        return Probe::electron;
+      } else if (probe == "neutron") {
+        return Probe::neutron;
+      }
+
+      DXTBX_ERROR("Unknown probe " + probe);
+      return Probe::xray;
```
Comment (ndevenish) [inline src/dxtbx/model/beam.h:339]:
```suggestion
      DXTBX_ERROR("Unknown probe " + probe);
```

Top line throws, right?

**Review by ndevenish — CHANGES_REQUESTED** [review]

**Review by ndevenish — APPROVED** [review]

**Pull request merged**